src/AutoConstructor.Generator/AutoConstructorGenerator.cs

```diff
@@ -23,142 +25,128 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             i.AddSource(Source.InjectAttributeFullName, SourceText.From(Source.InjectAttributeText, Encoding.UTF8));
         });
 
-        IncrementalValuesProvider<ClassDeclarationSyntax> classDeclarations = context.SyntaxProvider
-                .CreateSyntaxProvider(static (s, _) => IsSyntaxTargetForGeneration(s), static (ctx, _) => GetSemanticTargetForGeneration(ctx))
-                .Where(static m => m is not null)!;
-
-        IncrementalValueProvider<(Compilation compilation, ImmutableArray<ClassDeclarationSyntax> classes, AnalyzerConfigOptions options)> valueProvider =
-            context.CompilationProvider
-            .Combine(classDeclarations.Collect())
+        IncrementalValuesProvider<GeneratorExectutionResult?> valueProvider = context.SyntaxProvider
+            .ForAttributeWithMetadataName(
+                Source.AttributeFullName,
+                static (node, _) => IsSyntaxTargetForGeneration(node),
+                static (context, _) => (ClassDeclarationSyntax)context.TargetNode)
+            .Where(static m => m is not null)
             .Combine(context.AnalyzerConfigOptionsProvider.Select((c, _) => c.GlobalOptions))
-            .Select((c, _) => (compilation: c.Left.Left, classes: c.Left.Right, options: c.Right));
+            .Combine(context.CompilationProvider)
+            .Select((c, _) => Execute(c.Right, c.Left.Left, c.Left.Right));
 
-        context.RegisterSourceOutput(valueProvider, static (spc, source) => Execute(source.compilation, source.classes, spc, source.options));
+        context.RegisterSourceOutput(valueProvider, static (context, item) =>
+        {
+            if (item is not null)
+            {
+                if (!item.Diagnostics.IsEmpty)
```

Comment:
Avoid capturing diagnostics into your pipeline, if possible. Rather, have analyzers report all diagnostics, and simply do the same checks from your generator and skip generation (or only generate some empty/dummy stub) if the code is invalid. Otherwise this will also break incrementality (diagnostics are not incremental).